internal/gengapic/client_init.go

```diff
@@ -135,155 +135,236 @@ func (g *generator) clientOptions(serv *descriptor.ServiceDescriptorProto, servN
 	return nil
 }
 
-func (g *generator) clientInit(serv *descriptor.ServiceDescriptorProto, servName string) error {
+func (g *generator) abstractClientIntfInit(serv *descriptor.ServiceDescriptorProto, servName string) error {
 	p := g.printf
-
-	var hasLRO bool
+	p("// %sAbstractClient is an interface that defines the methods availaible from %s.", servName, g.apiName)
```

Comment:
Our approach for the Go client libraries thus far has been to avoid interfaces, especially public one. If a new method `Zap` got added this is technically a breaking change to the interface. I don't think the outside world needs to know about this though, so could we just make the interface private. Only the package itself would ever use this, right?

Reply:
nit: maybe just `fooClient` vs `fooAbstractClient`. All interfaces are abstract and the inclusion of the word feels strongly OOP/Java

Reply:
@vchudnov-g 

> wdym about the public interface? I think the idea here is for app developers to be able to code almost exclusively against the interface, after getting a concrete gRPC or REST client. So the interface needs to be public, right?

Why could they not keep coding against the concrete struct there is today and we just handle what transport is used under the hood? Like I said above, adding a method to an interface is a breaking change whereas adding one to a struct is not. This is why the clients did not provide an interface to begin with. In Go the general sentiment is if you want an interface for a struct you should declare it in your(the caller code). Also that functions should be returning concrete types and accepting interfaces. This is all done(mainly) for the sake of compatibility.

Reply:
You make a good point about adding a method being a technically breaking change, but IIUC this doesn't have serious impact in this particular instance. We're not expecting our users to define structs that implement the interface, and (both) structs in the generated library are structs that we control. If the API adds a `Zap` method, the gRPC client, the rest client, and the interface would all be updated simultaneously during regeneration.

Victor proposed a couple of usecases where application code does want the client/transport types to be polymorphic. Runtime transport choice was one, as well as dynamic fallback with one client of each kind. In both of these cases the user could definitely make their own interface to wrap both structs and obviate us from making this change. That feels like an unnecessary duplication of effort, though.

The workaround that Jean de Klerk mentions in [this post](https://blog.golang.org/module-compatibility) seems like overkill for our usecase here.

As for the `fooAbstractClient` naming convention, the `fooClient` moniker is already taken by the gRPC client. Reappropriating the name for the interface would itself be a breaking change.

The client struct as it is today is heavily gRPC centric. It seemed like it would be a lot of extra effort to partition off the gRPC fields, add rest fields, and maintain state that determined which kind of client our instance was. For Python GAPIC, the client receives (or creates) an injectable transport, but reworking that for golang GAPIC would be another breaking change.

Reply:
Personally, this feels like something we need to discuss in a doc/meeting. It doesn't feel like we are all on the same page with how the generated surface should change to support REGAPIC and how we can do that within best practices.

Reply:
I made a design doc a while ago describing alternatives and tradeoffs. I'll send it around for another round of scrutiny.

Reply:
I'm fine with meeting to discuss this. A couple of points to bring up:
- as Dov said, we're not expecting users to implement the interface, but just to use it. When our generators change the interface, they'll be changing its concrete implementations as well.
- it just occurred to me that maybe we also have the alternative of using embedded structs (I haven't thought this out fully). In that case, the abstract interface becomes an abstract struct with dummy methods that is embedded in the concrete structs. The concrete structs can define their own methods with the same names as the abstract struct, thus shadowing them. Any concrete method not such defined will fall back to the abstract struct's.  That said, I still prefer the interface approach, which seems simpler.

But yes, let's discuss! Certainly having the transport selection/use happen under the hood is another way to approach the same goal, and one that other languages have been using. I'm open to that as well; let's figure out what that could look like and whether it's preferable.